packages/protocol/contracts/adminnable/SelfAdminnable.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.6;
+
+import "./interfaces/ISelfAdminnable.sol";
+
+/// @title Contract that implements multiple levels of admins for multiple
+/// "adminned" addresses independently
+contract SelfAdminnable is ISelfAdminnable {
+    /// @notice Called to get the rank of an admin for the adminned address
+    mapping(address => mapping(address => uint256))
+        public
+        override adminnedToAdminToRank;
+
+    /// @dev Reverts if the caller's rank is not greater than or equal to
+    /// `rank` for the adminned address
+    /// @dev Adminned address always satisfies rank requirements
+    /// @param adminned Adminned address
+    /// @param rank Rank caller's rank will be compared to
+    modifier onlyWithRank(address adminned, uint256 rank) {
+        require(
+            msg.sender == adminned ||
+                adminnedToAdminToRank[adminned][msg.sender] >= rank,
+            "Caller ranked low"
+        );
+        _;
+    }
+
+    /// @notice Called by an admin of higher rank to set the rank of an admin
+    /// of lower rank for the adminned address
+    /// @param adminned Adminned address
+    /// @param targetAdmin Target admin address
+    /// @param newRank Rank to be set
+    function setRank(
+        address adminned,
+        address targetAdmin,
+        uint256 newRank
+    )
+        external
+        override
+        onlyWithRank(
+            adminned,
+            max(adminnedToAdminToRank[adminned][targetAdmin], newRank) + 1
+        )
+    {
+        adminnedToAdminToRank[adminned][targetAdmin] = newRank;
```

Comment:
Should we be preventing from setting `address(0)` on either level of the mapping?

Reply:
Admin ranks for zero `adminned` will all be zero so those calls will revert anyway. I added checks for zero `targetAdmin` https://github.com/api3dao/airnode/pull/556/commits/adee69da44fc5f93d0bfa97165221bdad02ededf